components/modules/rmt.c

```diff
@@ -0,0 +1,403 @@
+  }
+  free(p);
+}
+
+static void handle_receive(void *param) {
+  lrmt_cb_params *p = (lrmt_cb_params *) param;
+
+  RingbufHandle_t rb = NULL;
+
+  //get RMT RX ringbuffer
+  rmt_get_ringbuf_handle(p->channel, &rb);
+  // Start receive
+  rmt_rx_start(p->channel, true);
+  while (!p->rx_shutting_down) {
+    size_t length = 0;
+    rmt_item32_t *items = NULL;
+
+    items = (rmt_item32_t *) xRingbufferReceive(rb, &length, 50 / portTICK_PERIOD_MS);
+    if (items && length) {
+      lrmt_cb_params *rx_params = malloc(sizeof(lrmt_cb_params) + length);
+      if (rx_params) {
+        memset(rx_params, 0, sizeof(*rx_params));
+	memcpy(rx_params + 1, items, length);
+        rx_params->cb_ref = p->cb_ref;
+	rx_params->data = (void *) items;
+	rx_params->len = length;
+        rx_params->channel = p->channel;
+        task_post_high(cb_task_id, (task_param_t) rx_params);
+      } else {
+	printf("Unable allocate receive data memory\n");
+      }
+    }
+    if (items) {
+      vRingbufferReturnItem(rb, (void *) items);
+    }
+  }
+
+  p->dont_call = true;
+  task_post_high(cb_task_id, (task_param_t) p);
+
+  /* Destroy this task */
+  vTaskDelete(NULL);
+}
+
+static int lrmt_on(lua_State *L) {
+  lrmt_channel_t ud = (lrmt_channel_t)luaL_checkudata(L, 1, "rmt.channel");
+  if (ud->tx) {
+    return luaL_error(L, "Cannot receive on a TX channel");
+  }
+
+  luaL_argcheck(L, !strcmp(lua_tostring(L, 2), "data") , 2, "Must be 'data'");
+
+  luaL_argcheck(L, lua_type(L, 3) == LUA_TFUNCTION, 3, "Must be a function");
+
+  if (ud->rx_params) {
+    return luaL_error(L, "Can only call 'on' once");
+  }
+
+  // We have a callback
+  lrmt_cb_params *params = (lrmt_cb_params *) malloc(sizeof(*params));
+  if (!params) {
+    return luaL_error(L, "Cannot allocate memory");
+  }
+  memset(params, 0, sizeof(*params));
+
+  lua_pushvalue(L, 3);
+  params->cb_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+  ud->rx_params = params;
+  params->channel = ud->channel;
+
+  xTaskCreate(handle_receive, "rmt-rx-receiver", 3000, params, 2, NULL);
+
+  return 0;
+}
+
+static void wait_for_transmit(void *param) {
+  lrmt_cb_params *p = (lrmt_cb_params *) param;
+  esp_err_t rc = rmt_wait_tx_done(p->channel, 10000 / portTICK_PERIOD_MS);
+
+  p->rc = rc;
+  task_post_high(cb_task_id, (task_param_t) p);
+
+  /* Destroy this task */
+  vTaskDelete(NULL);
+}
+
+static int lrmt_send(lua_State *L) {
+  lrmt_channel_t ud = (lrmt_channel_t)luaL_checkudata(L, 1, "rmt.channel");
+  if (!ud->tx) {
+    return luaL_error(L, "Cannot send on an RX channel");
+  }
+  int string_index = 2;
+
+  if (lua_type(L, 2) == LUA_TTABLE) {
+    lua_getfield(L, 2, "concat");
+    lua_pushvalue(L, 2);
+    lua_pushstring(L, "");
+    lua_call(L, 2, 1);
+    string_index = -1;
+  }
+
+  size_t len;
+  const char *data = lua_tolstring(L, string_index, &len);
+  if (!data || !len) {
+    return 0;
+  }
+
+  if (len & 1) {
+    return luaL_error(L, "Length must be a multiple of 2");
+  }
+
+  if (len & 3) {
```

Comment:
Why `len & 3` here? You've filtered out `& 1` above already?
(Or am I having a senior's moment here?)

Reply:
Add a comment to explain